packages/core/src/getNextSnapshot.ts

```diff
@@ -0,0 +1,74 @@
+import { createEmptyActor } from './actors/index.ts';
+import {
+  ActorRefFrom,
+  ActorScope,
+  AnyActorLogic,
+  AnyActorScope,
+  EventFromLogic,
+  InputFrom,
+  SnapshotFrom
+} from './types.ts';
+
+/** @internal */
+export function createInertActorScope<T extends AnyActorLogic>(
+  _actorLogic: T
+): AnyActorScope {
+  const self = createEmptyActor() as ActorRefFrom<T>;
+  const inertActorScope: ActorScope<SnapshotFrom<T>, EventFromLogic<T>, any> = {
+    self,
+    defer: () => {},
+    id: '',
+    logger: () => {},
+    sessionId: '',
+    stopChild: () => {},
+    system: (self as any).system // TODO: fix types?
+  };
+
+  return inertActorScope;
+}
+
+export function getInitialSnapshot<T extends AnyActorLogic>(
+  actorLogic: T,
+  input: InputFrom<T>
+): SnapshotFrom<T> {
+  const actorScope = createInertActorScope(actorLogic);
+  return actorLogic.getInitialSnapshot(actorScope, input);
+}
+
+/**
+ * Determines the next snapshot for the given `actorLogic` based on
+ * the given `snapshot` and `event`.
+ *
+ * If the `snapshot` is `undefined`, the initial snapshot of the
+ * `actorLogic` is used.
+ *
+ * @example
+  ```ts
+  import { getNextSnapshot } from 'xstate';
+  import { trafficLightMachine } from './trafficLightMachine.ts';
+
+  const nextSnapshot = getNextSnapshot(
+    trafficLightMachine, // actor logic
+    undefined, // snapshot (or initial state if undefined)
+    { type: 'TIMER' }); // event object
+
+  console.log(nextSnapshot.value);
+  // => 'yellow'
+
+  const nextSnapshot2 = getNextSnapshot(
+    trafficLightMachine, // actor logic
+    nextSnapshot, // snapshot
+    { type: 'TIMER' }); // event object
+
+  console.log(nextSnapshot2.value);
+  // =>'red'
+  ```
+ */
+export function getNextSnapshot<T extends AnyActorLogic>(
+  actorLogic: T,
+  snapshot: SnapshotFrom<T>,
+  event: EventFromLogic<T>
+): SnapshotFrom<T> {
+  const inertActorScope = createInertActorScope(actorLogic);
+  return actorLogic.transition(snapshot, event, inertActorScope);
```

Comment:
if you do what you have mentioned [here](https://github.com/statelyai/xstate/pull/4596/files#r1432858780) then you kinda should do smth like this here:
```suggestion
  const inertActorScope = createInertActorScope(actorLogic);
  (inertActorScope.self as any)._snapshot = snapshot;
  return actorLogic.transition(snapshot, event, inertActorScope);
```

Reply:
otherwise the `self.getSnapshot()` will always return the initial snapshot in actions and that doesn't match how this would behave in a running machine

we also can't use `createActor(logic, { snapshot })` because this argument is meant to be the persisted snapshot... but we only have a snapshot here